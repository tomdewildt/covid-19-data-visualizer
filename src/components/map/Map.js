import React, { useState } from "react";
import DeckGL from "@deck.gl/react";
import { GeoJsonLayer } from "@deck.gl/layers";
import { HeatmapLayer } from "@deck.gl/aggregation-layers";
import { Map as MapGL } from "react-map-gl";
import PropTypes from "prop-types";

import Tooltip from "../tooltip/Tooltip";

const MAPBOX_TOKEN = process.env.REACT_APP_MAPBOX_API_TOKEN;
const MAPBOX_STYLE = process.env.REACT_APP_MAPBOX_STYLE;

const Map = ( { data, viewState } ) => {
    const [ tooltip, setTooltip ] = useState( {} );

    const layers = [
        new GeoJsonLayer( {
            data,
            opacity: 0.8,
            extruded: false,
            filled: true,
            stroked: true,
            pickable: true,
            getFillColor: [ 128, 128, 128, 0 ],
            getLineColor: [ 128, 128, 128, 255 ],
            lineWidthMinPixels: 1,
            lineWidthMaxPixel: 1,
            onHover: ( { x, y, object } ) => setTooltip( { x, y, object } ),
        } ),
        new HeatmapLayer( {
            data,
            getPosition: ( d ) => d.properties.Coordinaten,
            getWeight: ( d ) => d.properties.PositiefGetest,
            intensity: 1,
            treshold: 0.05,
            radiusPixels: 75,
        } ),
    ];

    return (
        <DeckGL
            controller
            layers={ layers }
            initialViewState={ viewState }
            onViewStateChange={ () => setTooltip( {} ) }
        >
            <MapGL
                reuseMaps
                preventStyleDiffing
                mapStyle={ MAPBOX_STYLE }
                mapboxApiAccessToken={ MAPBOX_TOKEN }
            />
            { tooltip.object && (
                <Tooltip
                    aria-labelledby="tooltip-title"
                    x={ tooltip.x }
                    y={ tooltip.y }
                >
                    <Tooltip.Title id="tooltip-title">
                        { tooltip.object.properties.Gemeente }
                    </Tooltip.Title>
                    <Tooltip.Text>
                        <b>Code: </b>
                        { tooltip.object.properties.Gemeentecode }
                    </Tooltip.Text>
                    <Tooltip.Text>
                        <b>Positief Getest: </b>
                        { tooltip.object.properties.PositiefGetest }
                    </Tooltip.Text>
                    <Tooltip.Text>
                        <b>Datum: </b>
                        { tooltip.object.properties.Datum }
                    </Tooltip.Text>
                </Tooltip>
            )}
        </DeckGL>
    );
};

Map.propTypes = {
    data: PropTypes.arrayOf( PropTypes.shape( {} ) ).isRequired,
    viewState: PropTypes.shape( {
        latitude: PropTypes.number.isRequired,
        longitude: PropTypes.number.isRequired,
        zoom: PropTypes.number.isRequired,
        pitch: PropTypes.number,
        bearing: PropTypes.number,
    } ).isRequired,
};

export default Map;
