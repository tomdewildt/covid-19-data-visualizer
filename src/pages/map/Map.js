import React, { useEffect } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { actions } from "./duck";
import {
    Page,
    Error,
    Loading,
    Map,
} from "../../components";

const PageMap = ( { dataset, error, LoadDataset } ) => {
    useEffect( () => {
        LoadDataset();
    }, [ LoadDataset ] );

    if ( error ) {
        return (
            <Page id="map">
                <Error message={ error } />
            </Page>
        );
    }
    if ( !dataset ) {
        return (
            <Page id="map">
                <Loading />
            </Page>
        );
    }

    const data = dataset.data.get( dataset.dates[ dataset.dates.length - 1 ] ) ?? [];
    return (
        <Page id="map">
            <Map
                data={ data }
                viewState={ {
                    latitude: 52.1326,
                    longitude: 5.2913,
                    zoom: 7,
                    pitch: 0,
                    bearing: 0,
                } }
            />
        </Page>
    );
};

PageMap.propTypes = {
    dataset: PropTypes.shape( {
        data: PropTypes.shape( {
            clear: PropTypes.func.isRequired,
            delete: PropTypes.func.isRequired,
            forEach: PropTypes.func.isRequired,
            get: PropTypes.func.isRequired,
            has: PropTypes.func.isRequired,
            set: PropTypes.func.isRequired,
            size: PropTypes.number.isRequired,
        } ).isRequired,
        dates: PropTypes.arrayOf( PropTypes.string ),
    } ),
    error: PropTypes.string,
    LoadDataset: PropTypes.func.isRequired,
};

PageMap.defaultProps = {
    dataset: null,
    error: null,
};

const mSTP = ( { map: { dataset, error } } ) => ( { dataset, error } );

const mDTP = ( dispatch ) => ( {
    LoadDataset: () => dispatch( actions.loadDataset() ),
} );

export default connect( mSTP, mDTP )( PageMap );
