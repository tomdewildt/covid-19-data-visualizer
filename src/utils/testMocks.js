/* eslint-disable react/prop-types */
import React from "react";

// Components
jest.mock( "@deck.gl/react", () => ( { children } ) => (
    <mock-deckgl>{ children }</mock-deckgl>
) );

// Assets
jest.mock( "../assets/gemeenten.json", () => ( [
    {
        type: "Feature",
        properties: { Gemeente: "Gemeente 1", Gemeentecode: 1, Coordinaten: [] },
        geometry: { type: "MultiPolygon", coordinates: [] },
    },
    {
        type: "Feature",
        properties: { Gemeente: "Gemeente 2", Gemeentecode: 2, Coordinaten: [] },
        geometry: { type: "MultiPolygon", coordinates: [] },
    },
    {
        type: "Feature",
        properties: { Gemeente: "Gemeente 3", Gemeentecode: 3, Coordinaten: [] },
        geometry: { type: "MultiPolygon", coordinates: [] },
    },
] ) );

