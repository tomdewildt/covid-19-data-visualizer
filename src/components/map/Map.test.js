import React from "react";
import { axe } from "jest-axe";

import { renderWithTheme } from "../../utils/testUtils";

import Map from "./Map";

//
// NOTE:
//
// This component needs better testing, unfortunatly this won't be possible
// until Deck.GL provides a better testing interface.
//
// - Test if layers are correctly generated
// - Test if the tooltip renders correctly on hover
// - Test if the tooltip is removed after hover
//
describe( "Map", () => {
    const data = [];
    const viewState = {
        latitude: 0.0,
        longitude: 0.0,
        zoom: 7,
        pitch: 0,
        bearing: 0,
    };

    it( "renders consistently", () => {
        const { asFragment } = renderWithTheme( (
            <Map
                data={ data }
                viewState={ viewState }
            />
        ) );
        const fragment = asFragment();

        expect( fragment ).toMatchSnapshot();
    } );

    it( "should have no axe violations", async () => {
        const { container } = renderWithTheme( (
            <Map
                data={ data }
                viewState={ viewState }
            />
        ) );
        const result = await axe( container );

        expect( result ).toHaveNoViolations();
    } );
} );
