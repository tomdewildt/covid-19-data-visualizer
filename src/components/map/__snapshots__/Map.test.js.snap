// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders consistently 1`] = `
<DocumentFragment>
  <mock-deckgl>
    <mock-mapgl />
  </mock-deckgl>
</DocumentFragment>
`;
